### DISCUSSION TITLE
A few Ideas and Suggestions

### DISCUSSION THREAD

**Discussion by YogiYang** [twinbasic/twinbasic#441] [Ideas]
Hello,

Lately I have been using a lot of other development tools right from Delphi to WinDev to FreeBasic to Lianja. And I have come to like a few features in some of these tools that I would like to suggest some here. I have not used tB extensively so if some of these suggestion have been implemented please ignore them...

- **For IDE** - Give a feature to find and list all reference of a Procedure or Function so we can know as from how many places a procedure or function is called.
- **For IDE** - Add facility for _Internal Libraries_ and _External Libraries_. Here users can put all the libraries/components that come with tB in _Internal Libraries_ folder and all third party libraries/components will have to be put in _External Libraries_ folder. This way it would be easy for us developers to transport the IDE along with all the dependencies from one PC to another PC by just coping the folder of VS Code. Here you have to take care that the IDE should not load all the libraries but instead allow user to select the libraries they want to use in a project and load only those libraries. This will prevent the IDE from being very sluggish at start up.
- **For Compiler** - Add facility to declare functions and procedure that can be called asynchronous (let us call it **_Async Subs_**) that is they will run in separate threads. There should be facility to pause the execution of such a thread (functions/procedure) until certain conditions are fulfilled or some value is returned from server or something like this. When these _Subs_ pause the rest of the application should not get stuck. For example I can write an asynchronous procedure to query availability of internet and when internet is available it should fire an event or call another routine to check for availability of new version of application, but all this should happen in separate thread that is running in the background without disturbing the main thread executing unless it is absolutely necessary to halt main thread and show users some information, etc.
- **For Both** - Give built in support for accessing SQLite DB. Here the IDE should provide features for creating/altering tables and their structures, entering data, etc. (If you are planning to integrate RC6 then support for SQLite is already there you will have to only build necessary UI for DB structure & data management)
- **For Both** - Add an Event driven reporting engine that developer can use for creating beautify reports.

A few built function suggestion that may prove to be productivity boosters for developers (These are just a few from top of my head):

- GetAppPath
- ExtractFileNameFromPath
- ExtractPath
- ExtractExtension
- ChangeExtension
- GetAppDrive
- GetListFoFilesInFolder (based on wildcard specified)
- PathExist
- FileExist
- CountWords (here user has to provide a delimited based on which words will be separated)
- LoopThroughEachWord (in string)
- FindFile(First/Next with support to search in sub folders)
- GetTempFolder
- GetWindowsFolder
- GetSystemFolder
- GetOSVersion
- ElevateToAdmin
- LockUIUpdate
- UnLockUIUpdate
- ShowFileOpenDialog
- ShowFolderSelectDialog
- ShowFileSaveDialog
- ShowPrintDialog
- GetFontList (This should be loadable in a StringList Object)
- IsNetworkFolder
- GetControlPanelFolder
- GetRecycleBinFolder
- GetDesktop
- GetDocumentFolder
- GetRoamingAppDataFolder
- GetProgramFilesX86Folder
- GetProgramFilesCommonX86Folder
- GetProgramFilesX64Folder
- GetProgramFilesCommonX64Folder
- Soundex
- SoundexWord
- DecodeSoundexWord
- SoundexSimilar
- SoundexCompare
- GetTLanguagesCount
- GetTLanguagesName
- GetTLanguagesNameFromLocaleID
- GetTLanguagesID
- GetTLanguagesLocaleID
- GetTLanguagesExt
- GetFileVersion
- GetCurrencyStringFromOS
- SetCurrencyStringInOS

A few Objects

- StreamObject (Memory/File)
- StringListObject (something like a listbox with facility to load/save content from/to file)
- IniFileObject (for handling create/read from INI or config files)
- RegistryObject
- RegExObject
- XMLObject
- PrinterObject
- SocketObject

**Comment by dmrvb** [comment]
Some good ideas.
For now I think the GUI side of things should be top priority, plus 64 bit and an optimised compiler. TB and IDE enhancements can come later.
Your list of built in functions is good but these mostly do not need to be built in (and should not be). It would be great if TB gets an active community like VB had, and many of these extensions can be produced by the community and are probably best provided as source code modules in TB (with a bit of API calling in some cases).
Quite a bit of this functionality is already available as VB6 code on the www, and once TB is up and running (with a GUI) I would be happy to spend some time on some of this code, though there are others here who can likely do a better job

**Comment by bclothier** [comment]
Just a comment. Those functions really need to have namespaces (see #111 ) to avoid bloating the intellisense. I would hate to have wade through at least ten keystrokes just to get to what I wanted. Namespace would also make it much more discoverable. I think it's more intuitive to find `Path.Exists()` than it is to do a `FileExists()` or `FolderExists()`.

Global namespace pollution is a _huge_ problem in JavaScript world. I don't want it to be in twinBasic, either. VBx already struggles with it, especially with implicit default members & appobjects crowding the global namespace. 

I agree that lot of those could be contributed by the community and with something similar to nugets (see #16 ) would make it easy for people to contribute & share, but before that is possible, we really do have to solve the matter of namespaces if we want to be able to organize the library functions so that they are easily discoverable.

**Reply by mburns08109** [reply]
Why can't namespaces be a user-definable thing?
I mean sure create a default namespace, but allow tB to load an XML file that defines/re-defines the namespaces.
That way a developer can customize the namspaces lookups to their greatest needs/fewest steps and let a more generic namespace tree fall under a "General" sub-category.

That seems like it'd be the best answer to any future "namespace bloat" issues.

**Reply by wqweto** [reply]
> That seems like it'd be the best answer to any future "namespace bloat" issues.

Do you mean to allow everyone to customize their namespaces by tweaking a bunch of XML files and loading them in the IDE?

Mind that this will make source code written under different XML files completely non-portable even from IDE to IDE so can't be a answer to anything IMO.

Or do you mean component authors to provide an XML file for their ActiveX library on the side?

Namespaces in other languages are intrinsic part of the source code identifiers as used in the source code. The problem is how to make namespaces compatible with COM type libraries. Can these be included in the TLB as hidden attributes for instance? Can namespaces be optional to be compatible with VBx?

**Reply by mburns08109** [reply]
Sorry - I didn't express my thoughts well.
I was thinking more along the lines of a user-defined ALIAS file. The actual underlying namespaces would be the same, but the local UI could prefer the ALIAS file (if present) to the original underlying names, but could/would absorb and use both.

**Reply by wqweto** [reply]
Again, this will make any source code written in one IDE incompatible to next user's IDE becuase their ALIAS files will differ.

That's why the typedefs are part of the source code so the source file begins with `using reader = MyLibrary.MyNamespace.FileReader` and this way the source file becomes "portable" accross projects/IDEs.

**Reply by mburns08109** [reply]
not if the code were always saved with the aliases expanded to their original libraries.
The aliases would only apply if the alias file were present.

**Reply by mansellan** [reply]
It's not just about pollution of the global namespace though, namespaces also avoid typename collisions between different libraries (`Excel.Range` vs `Word.Range` being a simple example). I wouldn't want something as fundamental as type location to be hived out to an external config file, it should be a first-class citizen of the language.

**Reply by wqweto** [reply]
Could be. Just curious which other language has this feature? Which IDE does something like this?

In C/C++ the typedefs can reside in a separate include files and this make source code hard to read and almost impossible to refactor via automation tools, all of this on top of slow compilation speeds.

**Reply by mburns08109** [reply]
I know of no IDE that does this now.
Think of this like the old MS-DOS SUBST command in effect.
It's a temporary reassignment of the original folderspec with a temporary drive letter.
The alias file idea is the same thing. The underlying items are what the compiler sees, just the IDE would do the "translating" between the alias'd things and the user's UI. no alias file in the IDE, no "translation" happens in the presentation layer of the IDE. the IDE always saves the underlying un-aliased library references.
